src/main/java/com/redhat/developers/msa/aloha/AlohaVerticle.java

```diff
@@ -56,36 +55,53 @@
-        } else {
-            // Brave configured for a Server
-            BRAVE = builder.spanCollector(HttpSpanCollector.create(System.getenv("ZIPKIN_SERVER_URL"),
-                new EmptySpanCollectorMetricsHandler()))
-                .build();
-        }
     }
 
     @Override
     public void start() throws Exception {
         Router router = Router.router(vertx);
         router.route().handler(ctx -> {
-            // note: this is *not* an example of how to properly integrate vert.x with zipkin
-            // for a more appropriate way to do that, check the vert.x documentation
-            ServerRequestInterceptor serverRequestInterceptor = BRAVE.serverRequestInterceptor();
-            serverRequestInterceptor.handle(new HttpServerRequestAdapter(new VertxHttpServerRequest(ctx.request()), new DefaultSpanNameProvider()));
-            ctx.data().put("zipkin.span", BRAVE.serverSpanThreadBinder().getCurrentServerSpan());
-            ctx.next();
-            ctx.addBodyEndHandler(v -> BRAVE.serverResponseInterceptor().handle(new HttpServerResponseAdapter(() -> ctx.response().getStatusCode())));
+            if (null != tracer) {
+                // first, we build or rebuild a context based on the request data
+                // there are two possible scenarios here: the user is calling this endpoint directly,
+                // or this endpoint is being called by another service, like /api/hola-chaining .
+                // If we are being called directly, this "extract" will not find any trace state and will create a new context
+                // from scratch.
+                // If we are being called by Hola, then we should have some HTTP headers with the trace state (trace ID), on which case
+                // we create a span context with that information.
+                SpanContext spanContext = tracer.extract(Format.Builtin.HTTP_HEADERS, new HttpHeadersExtractAdapter(ctx.request().headers()));
+
+                // now that we have the span context, we create a "child" span for this request
+                Span requestSpan = tracer.buildSpan(ctx.request().method().name())
+                        .asChildOf(spanContext)
+                        .start();
+
+                Tags.HTTP_URL.set(requestSpan, ctx.request().absoluteURI());
+                // we store the request span within the request data, for consumption on children spans that we might need
+                ctx.data().put(TRACING_REQUEST_SPAN, requestSpan);
+
+                // continue with the request processing
+                ctx.next();
+
+                // before the request finishes, we want to mark the request span as finished as well
+                ctx.addBodyEndHandler(v -> {
+                    Tags.HTTP_STATUS.set(requestSpan, ctx.response().getStatusCode());
+                    requestSpan.finish();
+                });
+            } else {
+                ctx.next();
```

Comment:
nit: I think this can be moved to out of `if else block..` and then removed on line 93 too.